server/registry/internal/system/artifacts.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package system
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/server/registry/internal/buildinfo"
+
+	"github.com/apigee/registry/server/registry/names"
+	"google.golang.org/genproto/googleapis/api/httpbody"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"gopkg.in/yaml.v3"
+)
+
+func Reserved(name names.Artifact) bool {
+	return name.ApiID() == "" && strings.HasPrefix(name.ArtifactID(), "system-")
+}
+
+func Artifact(name names.Artifact) (*rpc.Artifact, error) {
+	contents, mimeType, err := contents(name)
+	if err != nil {
+		return nil, err
+	}
+	message := &rpc.Artifact{
+		Name:      name.String(),
+		MimeType:  mimeType,
+		SizeBytes: int32(len(contents)),
+		Hash:      hashForBytes(contents),
+	}
+	return message, nil
+}
+
+func ArtifactContents(name names.Artifact) (*httpbody.HttpBody, error) {
+	contents, mimeType, err := contents(name)
+	if err != nil {
+		return nil, err
+	}
+	return &httpbody.HttpBody{
+		ContentType: mimeType,
+		Data:        contents,
+	}, nil
+}
+
+func contents(name names.Artifact) ([]byte, string, error) {
+	switch name.ArtifactID() {
+	case "system-buildinfo":
+		return system_buildinfo()
+	default:
+		return nil, "", status.Errorf(codes.NotFound, "%q not found", name)
+	}
+}
+
+func system_buildinfo() ([]byte, string, error) {
```

Comment:
Also, camelCase function names please.

Reply:
> ...camelCase...

Thanks, fixed.

> The return values seem unnecessarily generic. Can we return fully formed artifact messages instead? 

I prefer not to do that because I think it would create a pattern in which any function returning a system artifact would return a fully formed artifact and redundantly compute the content size and hash. Here there's a clear division of responsibility with content generation separate from artifact construction (and I am thinking that additional artifacts would be added, though I don't have any to suggest here).

> Are these system artifacts always going to be manifested (for lack of a better term) by the server at runtime? The implementation seems designed for that: allowing us to add more recognized IDs that generate a response.

Yes, exactly. This PR includes a "name grab" of the `system-` prefix to hold it in reserve for our use. I think we should clarify (I think I didn't make this clear enough) that system artifacts are experimental features that do not have the same guarantees of stability as the API surface. I think we can and should reserve the right to change system artifacts and their contents at any time (including removing them).

>> Would system- artifacts always be read-only? Maybe...

> This seems like it would be a pretty big change because it would introduce statefulness to system artifacts. I think that change would challenge a lot of the assumptions proposed in this PR, e.g. not including system artifacts in list operations. Maybe worth thinking a bit more about whether this will ever happen.

Indeed. I don't see that uncertainty blocking our usage of read-only system artifacts, but it is something that we should consider if we see a use for mutable system artifacts.

> Have we considered adding a collection specifically for system artifacts? Not one method per artifact as you two have been discussing, but also not special behavior for a subset of resources in the projects/*/locations/*/artifacts/* collection. Instead we would add an additional_binding on GetArtifact for a collection like projects/*/locations/*/systemArtifacts/*.

Yes, and I think this would be bad for the API design for the same reasons that we don't have application-specific collections of artifacts (there is no "apihubArtifacts" collection, for example). Codifying these subgroups would suggest that such new collections are expected or necessary when they aren't at all - simply by using nonoverlapping sets of ids, applications could share the artifact collection space and potentially benefit from sharing information among them -- though the biggest benefit is the reduction of friction for application developers. Focusing back on system artifacts, creating a "systemArtifacts" collection would also make it harder for us to change our mind about having system artifacts (because then we would also have to remove the binding from our protos) -- I think it's too soon to make that level of commitment.